File: src/lib/forecast.js
// The /api/forecast payload builder — contract v1.
//
// Every number a client renders is computed here, once, from the same
// src/lib/* modules the browser has always used. Nothing in this file is
// re-derivable by a client without reimplementing the maths, which is the
// whole point: see docs/forecast-api-contract.md and platform plan §2.
//
// Deliberately pure. No fetch, no process.env, no Date.now() except as a
// default argument — api/forecast.js does the I/O and hands the raw upstream
// body in. That keeps the payload testable against a fixed fixture and keeps
// the web-vs-endpoint agreement test honest.
import {
  LEVELS,
  RANGES,
  NOT_LINES,
  EPA_LINES,
  EPA_SENS,
  levelForPM25,
} from './rating.js';
import { ugm3ToAqi } from './aqi.js';
import { computeVerdict, verdictHeadline } from './verdict.js';
import { buildDaySummaries, buildPastDaySummaries } from './days.js';
import { trendAt } from './trend.js';
import { skyFor } from './sky.js';
import { applySensorAnchor, DECAY_HOURS } from './sensors.js';
import { computeAgreement, summarizeAgreement } from './agreement.js';
import { findNowIndex } from './openMeteo.js';
import { formatVerdictTime } from './time.js';

export const CONTRACT_VERSION = 1;
export const FORECAST_MODEL = 'cams-global';
export const PAST_DAYS = 3;
export const FORECAST_DAYS = 5;

const SOURCES = ['official', 'local', 'model'];

// Thrown for the cases the contract names in §9; api/forecast.js maps `code`
// straight onto the error envelope.
export class ForecastError extends Error {
  constructor(code, message) {
    super(message);
    this.name = 'ForecastError';
    this.code = code;
  }
}

export function normalizeSource(value) {
  return SOURCES.includes(value) ? value : 'official';
}

// Open-Meteo through /api/aq: one snapped point, timezone=auto. `auto` (rather
// than the UTC the grid fetch uses) is what carries the location's IANA zone
// and offset back, and every label in the payload is formatted in that zone —
// "when does it clear" is a question about the air over that place, not about
// the reader's laptop.
export function upstreamPath({ lat, lon }) {
  return (
    `/api/aq?latitude=${lat}&longitude=${lon}&hourly=pm2_5` +
    `&past_days=${PAST_DAYS}&forecast_days=${FORECAST_DAYS}&timezone=auto`
  );
}

const round = (v, places) =>
  v == null || Number.isNaN(v) ? null : Number(v.toFixed(places));
const round1 = (v) => round(v, 1);

// 'YYYY-MM-DDTHH:mm' — the naive-UTC form every src/lib function already
// speaks (they all do `new Date(t + 'Z')`). The payload adds the seconds and
// the Z on the way out.
function toNaiveUTC(localWallClock, utcOffsetSeconds) {
  const ms = Date.parse(`${localWallClock}Z`) - utcOffsetSeconds * 1000;
  if (Number.isNaN(ms)) throw new ForecastError('no-series', 'unparseable hourly time');
  return new Date(ms).toISOString().slice(0, 16);
}

const stampUTC = (naive) => `${naive}:00Z`;
const instantUTC = (ms) => new Date(ms).toISOString().replace(/\.\d{3}Z$/, 'Z');

// The rating ladder, shipped whole so no client hardcodes health copy —
// CLAUDE.md requires the explainer copy to ship verbatim, and copy retyped
// into a Swift file is copy that drifts.
export const SCALE = LEVELS.map((level, i) => ({
  index: level.index,
  key: level.key,
  name: level.name,
  rangeUg: RANGES[i],
  maxUg: level.max === Infinity ? null : level.max,
  visibility: level.visibility,
  notice: level.notice,
  notLine: NOT_LINES[i],
  guidance: { general: EPA_LINES[i], sensitive: EPA_SENS[i] },
}));

function officialRow(row) {
  if (!row || row.ug == null) return null;
  return {
    ug: round1(row.ug),
    aqi: ugm3ToAqi(row.ug),
    count: row.count ?? 0,
    area: row.area ?? null,
    distanceMi: row.distanceMi ?? null,
    // AirNow's own local wall-clock stamp, no zone. Passed through only when
    // it looks like one — inventing a zone would be worse than dropping it.
    observedAt: /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}$/.test(row.observedAt ?? '')
      ? row.observedAt
      : null,
  };
}

function localRow(row) {
  if (!row || row.ug == null) return null;
  return {
    ug: round1(row.ug),
    aqi: ugm3ToAqi(row.ug),
    count: row.count ?? 0,
    medianDistanceMi: row.medianDistanceMi ?? null,
  };
}

function skyPayload(pm25, naiveTime, lat, lon) {
  // An hour with no model value still has a sun: compute the sky at 0 µg/m³
  // rather than dropping it, so a gap darkens nothing and hides nothing.
  const s = skyFor(pm25 ?? 0, new Date(`${naiveTime}Z`), lat, lon);
  return {
    zenith: s.zenithRGB,
    mid: s.midRGB,
    horizon: s.horizonRGB,
    isDark: s.isDark,
    starOpacity: round(s.starOpacity, 3),
    smoke: { s1: round(s.smoke.s1, 3), s2: round(s.smoke.s2, 3) },
    sun: {
      altitudeDeg: round(s.sun.altitudeDeg, 2),
      azimuthDeg: round(s.sun.azimuthDeg, 2),
      visible: s.sun.visible,
      xFrac: round(s.sun.xFrac, 3),
      yFrac: round(s.sun.yFrac, 3),
      dim: round(s.sun.dim, 3),
    },
    moon: {
      altitudeDeg: round(s.moon.altitudeDeg, 2),
      azimuthDeg: round(s.moon.azimuthDeg, 2),
      visible: s.moon.visible,
      xFrac: round(s.moon.xFrac, 3),
      yFrac: round(s.moon.yFrac, 3),
      phaseFraction: round(s.moon.phaseFraction, 4),
    },
  };
}

// Which measured row anchors the delivered series. Degrades
// official -> local -> model (or local -> official -> model when asked for
// local), and falls all the way to model whenever the anchor would be a
// no-op, so the contract's "applied === 'model' implies pm25 === pm25Model"
// promise holds.
function resolveAnchor(requested, measured, modelNowUg) {
  const preference =
    requested === 'model' ? [] : requested === 'local' ? ['local', 'official'] : ['official', 'local'];
  if (modelNowUg == null) return { applied: 'model', ug: null };
  for (const key of preference) {
    const ug = measured?.[key]?.ug;
    if (ug != null) return { applied: key, ug };
  }
  return { applied: 'model', ug: null };
}

/**
 * Build the full contract-v1 payload.
 *
 * @param {object}  args
 * @param {object}  args.raw        Open-Meteo body as returned through /api/aq
 *                                  (timezone=auto: hourly.time is local wall
 *                                  clock, utc_offset_seconds carries the zone).
 * @param {{lat,lon}} args.requested Coordinates as the client sent them.
 * @param {{lat,lon}} args.snapped   Coordinates actually fetched (grid.js lattice).
 * @param {object|null} args.measured `{ official, local }` from /api/sensors, or null.
 * @param {string}  args.requestedSource 'official' | 'local' | 'model'.
 * @param {number}  args.nowMs      The instant to treat as "now".
 */
export function buildForecastPayload({
  raw,
  requested,
  snapped,
  measured = null,
  requestedSource = 'official',
  nowMs = Date.now(),
}) {
  const localTimes = raw?.hourly?.time;
  const modelSeries = raw?.hourly?.pm2_5;
  if (!Array.isArray(localTimes) || !localTimes.length || !Array.isArray(modelSeries)) {
    throw new ForecastError('no-series', 'upstream returned no hourly series');
  }

  const utcOffsetSeconds = Number.isFinite(raw.utc_offset_seconds) ? raw.utc_offset_seconds : 0;
  const timezone = raw.timezone || 'UTC';
  const timesUTC = localTimes.map((t) => toNaiveUTC(t, utcOffsetSeconds));
  const nowIndex = findNowIndex(timesUTC, nowMs);

  const pm25Model = modelSeries.map((v) => (v == null || Number.isNaN(v) ? null : v));
  const source = normalizeSource(requestedSource);
  const anchor = resolveAnchor(source, measured, pm25Model[nowIndex]);
  const pm25 = applySensorAnchor(pm25Model, nowIndex, anchor.ug);
  const offsetUg = anchor.ug == null ? 0 : round1(anchor.ug - pm25Model[nowIndex]);

  const verdict = computeVerdict({ pm25, nowIndex });
  const label = (i) => (i == null ? null : formatVerdictTime(timesUTC[i], timezone));
  const at = (i) => (i == null ? null : stampUTC(timesUTC[i]));

  const agreementRows = computeAgreement({
    timesUTC,
    pm25: pm25Model, // the band compares models, so it reads the un-anchored series
    fetchedAtMs: nowMs,
    hrrrSeries: null, // v1 is lead-time fade only — see contract §8
  });
  const agreementSummary = summarizeAgreement(agreementRows, { multiModel: false });

  const hours = timesUTC.map((t, i) => ({
    t: stampUTC(t),
    pm25: round1(pm25[i]),
    pm25Model: round1(pm25Model[i]),
    aqi: ugm3ToAqi(pm25[i]),
    levelIndex: levelForPM25(pm25[i])?.index ?? null,
    trend: pm25[i] == null ? null : trendAt(pm25, i, verdict),
    agreement: agreementRows[i].status,
    sky: skyPayload(pm25[i], t, snapped.lat, snapped.lon),
  }));

  const daySummaries = buildDaySummaries({ timesUTC, pm25, nowIndex, timezone });
  const pastSummaries = buildPastDaySummaries({ timesUTC, pm25, nowIndex, timezone });

  return {
    v: CONTRACT_VERSION,
    generatedAt: instantUTC(nowMs),

    location: {
      requested: { lat: requested.lat, lon: requested.lon },
      snapped: { lat: snapped.lat, lon: snapped.lon },
      timezone,
      utcOffsetSeconds,
    },

    now: {
      index: nowIndex,
      timeUTC: stampUTC(timesUTC[nowIndex]),
      exactUTC: instantUTC(nowMs),
    },

    window: { pastHours: PAST_DAYS * 24, forecastHours: FORECAST_DAYS * 24 },

    source: { requested: source, applied: anchor.applied, model: FORECAST_MODEL },

    scale: SCALE,
    hours,

    verdict: {
      above: verdict.above,
      levelIndex: verdict.nowLevelIndex,
      trend: verdict.trend,
      headline: verdictHeadline(verdict, label),

      clearIndex: verdict.clearIdx,
      clearAtUTC: at(verdict.clearIdx),
      clearLabel: label(verdict.clearIdx),

      arrivalIndex: verdict.arrivalIdx,
      arrivalAtUTC: at(verdict.arrivalIdx),
      arrivalLabel: label(verdict.arrivalIdx),

      peakIndex: verdict.peakIdx,
      peakAtUTC: stampUTC(timesUTC[verdict.peakIdx]),
      peakPm25: round1(pm25[verdict.peakIdx]),
    },

    days: daySummaries.map((d) => ({
      key: d.key,
      weekday: d.weekday,
      // buildDaySummaries falls back to level 0 for an all-null day; the
      // contract says a day with nothing in it has no level.
      levelIndex: d.max == null ? null : (d.level?.index ?? null),
      minPm25: round1(d.min),
      maxPm25: round1(d.max),
      dayParts: d.dayParts.map((p) => ({
        key: p.key,
        label: p.label,
        bucket: p.bucket ? { name: p.bucket.name, color: p.bucket.color } : null,
      })),
    })),

    pastDays: pastSummaries.map((d) => ({
      key: d.key,
      weekday: d.weekday,
      levelIndex: d.level?.index ?? null,
      minPm25: round1(d.min),
      maxPm25: round1(d.max),
    })),

    measured: {
      official: officialRow(measured?.official),
      local: localRow(measured?.local),
      model: {
        ug: round1(pm25Model[nowIndex]),
        aqi: ugm3ToAqi(pm25Model[nowIndex]),
      },
      anchor: { source: anchor.applied, offsetUg, decayHours: DECAY_HOURS },
    },

    agreement: {
      multiModel: false,
      diverged: agreementSummary.diverged,
      label: agreementSummary.label,
    },
  };
}

File: apple/Sources/SmokeshowKit/Contract/Forecast.swift
// The /api/forecast contract v1, as Swift.
//
// This file is a *decoder*, not a model of the domain. Every number in it was
// computed on the edge by src/lib/forecast.js. Nothing here derives a verdict,
// a clear-time, a rating threshold, or a trend — that is the entire reason the
// endpoint exists (docs/forecast-api-contract.md §6, platform plan §2). A
// native app that recomputes a clear-time from `hours` is a bug even when it
// agrees.
//
// Decoding rules the contract imposes, all implemented here:
//   • unknown fields are ignored, never rejected (additive changes ship
//     without a version bump);
//   • unknown *enum cases* decode to `.unknown` rather than throwing;
//   • every documented null is an Optional — a null is not an error;
//   • `v != 1` is "unavailable", not a partial parse (see ForecastEnvelope).

import Foundation

// MARK: - Envelope

/// Either a forecast or the error shape. The endpoint uses one envelope for
/// both so a decoder never sees two shapes (contract §9).
public enum ForecastEnvelope: Sendable {
    case forecast(Forecast)
    case failure(ForecastAPIError)
}

public struct ForecastAPIError: Decodable, Sendable, Equatable {
    public enum Code: String, Decodable, Sendable {
        case badCoords = "bad-coords"
        case upstreamFailed = "upstream-failed"
        case noSeries = "no-series"
        case internalError = "internal"
        case unknown

        public init(from decoder: Decoder) throws {
            let raw = try decoder.singleValueContainer().decode(String.self)
            self = Code(rawValue: raw) ?? .unknown
        }
    }

    public let code: Code
    public let message: String
}

// MARK: - Forecast

public struct Forecast: Decodable, Sendable, Equatable {

    /// The only version this build understands. A payload with any other `v`
    /// must be treated as unavailable — see `ForecastDecoder`.
    public static let supportedVersion = 1

    public let v: Int
    public let generatedAt: Date
    public let location: Location
    public let now: Now
    public let window: Window
    public let source: SourceInfo
    public let scale: [ScaleEntry]
    public let hours: [Hour]
    public let verdict: Verdict
    public let days: [Day]
    public let pastDays: [PastDay]
    public let measured: Measured
    public let agreement: AgreementSummary

    // MARK: Location

    public struct Coordinates: Decodable, Sendable, Equatable {
        public let lat: Double
        public let lon: Double
    }

    public struct Location: Decodable, Sendable, Equatable {
        /// What the client sent. Clients send *real* coordinates and read
        /// `snapped` back; pre-snapping forks a future lattice change.
        public let requested: Coordinates
        /// What was actually fetched — the 0.1° lattice from src/lib/grid.js.
        public let snapped: Coordinates
        /// IANA zone. Never null; "UTC" is the degenerate fallback.
        public let timezone: String
        /// Offset in effect at `now` *only*. Not valid across a DST boundary —
        /// use `timezone` with the platform zone database for arbitrary hours.
        public let utcOffsetSeconds: Int

        /// The zone every label in this payload was formatted in.
        public var timeZone: TimeZone {
            TimeZone(identifier: timezone) ?? TimeZone(secondsFromGMT: utcOffsetSeconds) ?? .gmt
        }
    }

    // MARK: Now

    public struct Now: Decodable, Sendable, Equatable {
        /// Always a valid index into `hours`. Accurate to ±30 minutes by
        /// construction, and may be one bucket stale from the 10-minute cache.
        public let index: Int
        public let timeUTC: Date
        /// The instant the server used.
        public let exactUTC: Date
    }

    public struct Window: Decodable, Sendable, Equatable {
        public let pastHours: Int
        public let forecastHours: Int
    }

    // MARK: Source

    public enum MeasurementSource: String, Decodable, Sendable, Equatable {
        case official, local, model
        case unknown

        public init(from decoder: Decoder) throws {
            let raw = try decoder.singleValueContainer().decode(String.self)
            self = MeasurementSource(rawValue: raw) ?? .unknown
        }
    }

    public struct SourceInfo: Decodable, Sendable, Equatable {
        public let requested: MeasurementSource
        /// Degrades official → local → model when the requested row is absent.
        public let applied: MeasurementSource
        /// The forecast model behind `pm25Model`, e.g. "cams-global".
        public let model: String
    }

    // MARK: Scale

    /// One rung of the rating ladder. All five ship in every response so that
    /// no client hardcodes health copy — CLAUDE.md requires the disclaimer and
    /// explainer copy to ship verbatim, and copy pasted into a Swift file is
    /// copy that drifts.
    public struct ScaleEntry: Decodable, Sendable, Equatable, Identifiable {
        public enum Key: String, Decodable, Sendable, Equatable {
            case allClear = "all-clear"
            case something
            case smells
            case tastes
            case smokeshow
            case unknown

            public init(from decoder: Decoder) throws {
                let raw = try decoder.singleValueContainer().decode(String.self)
                self = Key(rawValue: raw) ?? .unknown
            }
        }

        public struct Guidance: Decodable, Sendable, Equatable {
            public let general: String
            /// One rung stricter, for sensitive households. Which one a client
            /// shows is a local preference; both always ship.
            public let sensitive: String
        }

        public let index: Int
        public let key: Key
        public let name: String
        public let rangeUg: String
        /// Null on index 4 only (unbounded).
        public let maxUg: Double?
        public let visibility: String
        public let notice: String
        public let notLine: String
        public let guidance: Guidance

        public var id: Int { index }
    }

    // MARK: Hours

    public enum HourTrend: String, Decodable, Sendable, Equatable {
        case rising, falling, steady
        case unknown

        public init(from decoder: Decoder) throws {
            let raw = try decoder.singleValueContainer().decode(String.self)
            self = HourTrend(rawValue: raw) ?? .unknown
        }
    }

    public enum Agreement: String, Decodable, Sendable, Equatable {
        case agree, fade, diverge
        case unknown

        public init(from decoder: Decoder) throws {
            let raw = try decoder.singleValueContainer().decode(String.self)
            self = Agreement(rawValue: raw) ?? .unknown
        }
    }

    public struct Hour: Decodable, Sendable, Equatable {
        public let t: Date
        /// Delivered series: model plus the measured anchor. **Null is real.**
        /// Render an unknown hour, never 0 — 0 µg/m³ is a claim about clean
        /// air and would be a lie (contract §4).
        public let pm25: Double?
        public let pm25Model: Double?
        public let aqi: Int?
        /// Index into `scale`. Null exactly when `pm25` is null.
        public let levelIndex: Int?
        public let trend: HourTrend?
        public let agreement: Agreement
        /// Never null in v1; Optional so a future `?sky=off` can omit it.
        public let sky: Sky?

        /// True when the model has no value for this hour.
        public var isGap: Bool { pm25 == nil }
    }

    // MARK: Sky

    public struct Sky: Decodable, Sendable, Equatable {
        public struct Smoke: Decodable, Sendable, Equatable {
            public let s1: Double
            public let s2: Double
        }

        public struct Sun: Decodable, Sendable, Equatable {
            public let altitudeDeg: Double
            public let azimuthDeg: Double
            public let visible: Bool
            /// 0 = left/east … 1 = right/west.
            public let xFrac: Double
            /// 0 = top … 1 = bottom.
            public let yFrac: Double
            /// 0 clear → 1 smoke-dimmed.
            public let dim: Double
        }

        /// The moon, computed on the edge next to the sun so every client paints
        /// the identical phase and position rather than each running its own
        /// ephemeris. Emitted even below the horizon (negative altitude) — the
        /// renderer fades it in as it rises.
        public struct Moon: Decodable, Sendable, Equatable {
            public let altitudeDeg: Double
            public let azimuthDeg: Double
            public let visible: Bool
            /// 0 = left/east … 1 = right/west.
            public let xFrac: Double
            /// 0 = top … 1 = bottom.
            public let yFrac: Double
            /// 0 new … 0.5 full … 1 new again.
            public let phaseFraction: Double
        }

        public let zenith: RGB
        public let mid: RGB
        public let horizon: RGB
        /// The ink inversion. Paint the foreground light when true; do not
        /// re-derive it from the colours.
        public let isDark: Bool
        public let starOpacity: Double
        public let smoke: Smoke
        public let sun: Sun
        public let moon: Moon
    }

    /// `[r, g, b]`, each 0–255. Integer triples, not CSS strings, so every
    /// platform builds its own colour type without parsing.
    public struct RGB: Decodable, Sendable, Equatable {
        public let r: Int
        public let g: Int
        public let b: Int

        public init(r: Int, g: Int, b: Int) {
            self.r = r
            self.g = g
            self.b = b
        }

        public init(from decoder: Decoder) throws {
            var container = try decoder.unkeyedContainer()
            r = try container.decode(Int.self)
            g = try container.decode(Int.self)
            b = try container.decode(Int.self)
        }
    }

    // MARK: Verdict

    /// The answer. Every field here was computed once, on the edge. Render
    /// `headline`, `clearLabel`, and `arrivalLabel` verbatim.
    public struct Verdict: Decodable, Sendable, Equatable {
        public enum Trend: String, Decodable, Sendable, Equatable {
            case clearing, stuck, worsening, steady
            case unknown

            public init(from decoder: Decoder) throws {
                let raw = try decoder.singleValueContainer().decode(String.self)
                self = Trend(rawValue: raw) ?? .unknown
            }
        }

        /// PM2.5 at `now` is ≥ 35 µg/m³ ("Hazy").
        public let above: Bool
        public let levelIndex: Int
        public let trend: Trend
        /// One of exactly five sentences. Render as-is: it is the one string
        /// guaranteed identical on a user's phone and their laptop.
        public let headline: String

        /// First hour below 35 µg/m³ that holds for 6. Null together with its
        /// siblings when `above` is false or no sustained clear exists.
        public let clearIndex: Int?
        public let clearAtUTC: Date?
        /// "Thursday ~6 PM". The tilde is required — the forecast is an
        /// estimate and the label must say so.
        public let clearLabel: String?

        /// First hour at or above 35 µg/m³ that holds for 3.
        public let arrivalIndex: Int?
        public let arrivalAtUTC: Date?
        public let arrivalLabel: String?

        /// Never null — worst case the peak is `now` itself.
        public let peakIndex: Int
        public let peakAtUTC: Date
        public let peakPm25: Double?
    }

    // MARK: Days

    public struct DayPart: Decodable, Sendable, Equatable, Identifiable {
        public enum Key: String, Decodable, Sendable, Equatable {
            case morning, afternoon, evening
            case unknown

            public init(from decoder: Decoder) throws {
                let raw = try decoder.singleValueContainer().decode(String.self)
                self = Key(rawValue: raw) ?? .unknown
            }
        }

        /// The coarse 4-step strip scale — *not* the 5-level rating scale.
        /// Do not cross-index them.
        public struct Bucket: Decodable, Sendable, Equatable {
            public let name: String
            /// `#rrggbb`, straight from the server.
            public let color: String
        }

        public let key: Key
        public let label: String
        /// Null when no hour of the day fell in this part — commonly today's
        /// already-elapsed parts.
        public let bucket: Bucket?

        public var id: String { key.rawValue }
    }

    public struct Day: Decodable, Sendable, Equatable, Identifiable {
        /// Local calendar date, `YYYY-MM-DD`, in `location.timezone`.
        public let key: String
        public let weekday: String
        /// Level of the day's *worst* hour. Null when every hour is null.
        public let levelIndex: Int?
        public let minPm25: Double?
        public let maxPm25: Double?
        /// Always exactly 3, in morning/afternoon/evening order.
        public let dayParts: [DayPart]

        public var id: String { key }
    }

    /// Model estimate, never observation. A client that shows these must say
    /// so — `Copy.modelEstimate` exists for exactly this.
    public struct PastDay: Decodable, Sendable, Equatable, Identifiable {
        public let key: String
        public let weekday: String
        public let levelIndex: Int?
        public let minPm25: Double?
        public let maxPm25: Double?

        public var id: String { key }
    }

    // MARK: Measured

    /// Official, local, model. **Never averaged** — during fast-moving smoke a
    /// regulatory monitor 38 miles away and a cluster of consumer sensors 8
    /// miles away legitimately disagree, and blending them yields a number
    /// neither source said.
    public struct Measured: Decodable, Sendable, Equatable {
        public struct Row: Decodable, Sendable, Equatable {
            public let ug: Double
            public let aqi: Int
            public let count: Int
            public let area: String?
            public let distanceMi: Double?
            public let medianDistanceMi: Double?
            /// AirNow's own local wall-clock stamp — no zone, no `Z`.
            /// Display-only. Do not parse it into an instant.
            public let observedAt: String?
        }

        public struct ModelRow: Decodable, Sendable, Equatable {
            public let ug: Double?
            public let aqi: Int?
        }

        /// What was done to the model series to produce `hours[].pm25`.
        public struct Anchor: Decodable, Sendable, Equatable {
            public let source: MeasurementSource
            public let offsetUg: Double
            public let decayHours: Double
        }

        public let official: Row?
        public let local: Row?
        public let model: ModelRow
        /// Never null. `.model` with `offsetUg == 0` means nothing measured
        /// was available.
        public let anchor: Anchor
    }

    // MARK: Agreement

    public struct AgreementSummary: Decodable, Sendable, Equatable {
        /// v1 always false: the endpoint compares one model against lead time,
        /// not against a second model.
        public let multiModel: Bool
        public let diverged: Bool
        /// Display copy — render this string.
        public let label: String
    }
}
